main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala

```diff
@@ -123,69 +129,160 @@ object EffectBinder {
-      Expr.TryWith(exp, effUse, rules, tpe, purity, loc)
+      val e = letBindEffectsTopLevel(exp)
+      val rs = rules.map {
+        case HandlerRule(op, fparams, handlerExp) => HandlerRule(op, fparams, letBindEffectsTopLevel(handlerExp))
+      }
+      Expr.TryWith(e, effUse, rs, tpe, purity, loc)
 
     case Expr.Do(op, exps, tpe, purity, loc) =>
-      Expr.Do(op, exps, tpe, purity, loc)
+      val es = exps.map(letBindEffects(binders))
+      Expr.Do(op, es, tpe, purity, loc)
 
     case Expr.NewObject(name, clazz, tpe, purity, methods, exps, loc) =>
-      Expr.NewObject(name, clazz, tpe, purity, methods, exps, loc)
+      val es = exps.map(letBindInnerEffects(binders))
+      Expr.NewObject(name, clazz, tpe, purity, methods, es, loc)
 
     case Expr.Resume(exp, tpe, loc) =>
-      Expr.Resume(exp, tpe, loc)
+      val e = letBindEffects(binders)(exp)
+      Expr.Resume(e, tpe, loc)
+  }
+
+  /**
+    * Let-binds all sub-expressions including the top-most one.
+    * Always returns a variable or a constant.
+    */
+  private def letBindEffects(binders: mutable.ArrayBuffer[Expr])(exp: Expr)(implicit lctx: LocalContext, flix: Flix): Expr = {
+    val expTransformed = letBindInnerEffects(binders)(exp)
+    def bind(e: Expr): Expr = e match {
+      // trivial expressions
+      case Expr.Cst(_, _, _) => e
+      case Expr.Var(_, _, _) => e
+      case Expr.JumpTo(_, _, _, _) => e
+      case Expr.ApplyAtomic(_, _, _, _, _) => e
+      // non-trivial expressions
+      case Expr.ApplyClo(_, _, _, _, _, _) => letBindExpr(binders)(e)
+      case Expr.ApplyDef(_, _, _, _, _, _) => letBindExpr(binders)(e)
+      case Expr.ApplySelfTail(_, _, _, _, _, _) => letBindExpr(binders)(e)
+      case Expr.IfThenElse(_, _, _, _, _, _) => letBindExpr(binders)(e)
+      case Expr.Branch(_, _, _, _, _) => letBindExpr(binders)(e)
+      case Expr.Let(sym, exp1, exp2, _, _, loc) =>
+        binders.addOne(Expr.Let(sym, exp1, null, null, null, loc))
+        bind(exp2)
+      case Expr.LetRec(varSym, index, defSym, exp1, exp2, _, _, loc) =>
+        binders.addOne(Expr.LetRec(varSym, index, defSym, exp1, null, null, null, loc))
+        bind(exp2)
+      case Expr.Scope(_, _, _, _, _) => letBindExpr(binders)(e)
+      case Expr.TryCatch(_, _, _, _, _) => letBindExpr(binders)(e)
+      case Expr.TryWith(_, _, _, _, _, _) => letBindExpr(binders)(e)
+      case Expr.Do(_, _, _, _, _) => letBindExpr(binders)(e)
+      case Expr.Resume(_, _, _) => letBindExpr(binders)(e)
+      case Expr.NewObject(_, _, _, _, _, _, _) => letBindExpr(binders)(e)
+    }
+    bind(expTransformed)
+  }
+
+  private def letBindExpr(binders: mutable.ArrayBuffer[Expr])(e: Expr)(implicit lctx: LocalContext, flix: Flix): Expr.Var = {
+    val loc = e.loc.asSynthetic
+    val sym = Symbol.freshVarSym("anf", BoundBy.Let, loc)(Level.Default, flix)
+    lctx.lparams.addOne(LocalParam(sym, e.tpe))
+    binders.addOne(Expr.Let(sym, e, null, null, null, loc))
```

Comment:
What are these null values? Isn't that a disaster waiting to happen?

Reply:
I assume they only exist temporarily, but could we have some documentation + examples + a helper function that creates them and explains?

Reply:
It seems to me  you could create your own type:

```
Binder = LetBinder | LetRecBinder | LocalParamBinder
```

to avoid this?

Reply:
done